--- check_mailman_qfiles.pl	2009-08-04 08:34:39.246110511 +0300
+++ check_mailman_qfiles	2010-08-25 17:34:49.251923002 +0300
@@ -13,6 +13,7 @@
 use Getopt::Long;
 
 my $qfiles_base = '/var/spool/mailman';
+my $qfiles_data = '/var/lib/mailman/data';
 my %opts = (
 	warning => 5,   # 5 minutes
 	critical => 20, # 20 minutes
@@ -54,6 +55,29 @@
 	}
 }
 
+# check moderation queue
+my %modqueue;
+foreach (File::Find::Rule->file->name('heldmsg-*.pck')->in($qfiles_data)) {
+	if (my($list, $id) = /heldmsg-(.+)-(\d+)\.pck/) {
+		$modqueue{$list}++;
+	}
+}
+while (my($list, $count) = each %modqueue) {
+	if ($count > $opts{critical}) {
+		$problems{$list} = [ 'CRITICAL' ];
+		$problem_status = 'CRITICAL';
+	}
+	elsif ($count > $opts{warning}) {
+		$problems{$list} = [ 'WARNING' ];
+		$problem_status = 'WARNING' if ! $problem_status;;
+	}
+
+	if ($problems{$list}) {
+		$problems{$list}[1] = sprintf("%d moderation tasks", $count);
+	}
+}
+
+
 if (! $problem_status) {
 	print "OK: all normal\n";
 	exit 0;
